docs/design/core/waiters/README.md

```diff
@@ -0,0 +1,447 @@
+     *
+     * @param errorPredicate
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> errorAcceptor(Predicate<Throwable> errorPredicate) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.FAILURE;
+            }
+
+            @Override
+            public boolean matches(Throwable t) {
+                return errorPredicate.test(t);
+            }
+        };
+    }
+
+    /**
+     * Creates a retry waiter acceptor which determines if the exception should transition the waiter to retry state
+     *
+     * @param errorPredicate
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> retryAcceptor(Predicate<Throwable> errorPredicate) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.RETRY;
+            }
+
+            @Override
+            public boolean matches(Throwable t) {
+                return errorPredicate.test(t);
+            }
+        };
+    }
+}
+```
+
+## FAQ
+
+### For which services will we generate waiters?
+
+We will generate a `{Service}Waiter` class if the service has any operations that need waiter support.
+
+### Why not create waiter operations directly on the client?
+
+The options are: (1) create separate waiter utility classes or (2) create waiter operations on the client
+
+The following compares Option 1 to Option 2, in the interest of illustrating why Option 1 was chosen.
+
+**Option 1:** create separate waiter utility classes
+
+```Java
+dynamodb.waiter().waitUtilTableExists(describeTableRequest)
+```
+
+**Option 2:** create waiter operations on each service client
+
+```Java
+dynamodb.waitUntilTableExists(describeTableRequest)
+```
+
+**Option 1 Pros:**
+
+1. consistent with existing s3 utilities and presigner method approach, eg: s3Client.utilities()
+2. similar api to v1 waiter, and it might be easier for customers who are already using v1 waiter to migrate to v2.
+
+**Option 2 Pros:**
+
+1. slightly better discoverability
+
+**Decision:** Option 1 will be used, because it is consistent with existing features and option2 might bloat the size
+of the client, making it more difficult to use.
+
+### Why returning the service response for waiter operations with a success state of a specific response?
+
+The reason that the last response that has satisfied the waiter success state is returned is because it is a common pattern that customers creates a resource and then
+retrieves the information of the resources. Without returning the response, customers will have to send an extra request when the waiter returns. 
+See [feature request](https://github.com/aws/aws-sdk-java/issues/815)
```

Comment:
Can we cache *every* response, so they can see what happened, instead of just the latest?